Implement the delete function in the Cookie Store API

RupinMittal opened the pull request:
#### 525b38fe412e28097b9c30af5d549d13bd19899c
<pre>
Implement the delete function in the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259220">https://bugs.webkit.org/show_bug.cgi?id=259220</a>

Reviewed by Chris Dumez.

The delete function creates a cookie with the options given and
sets the expires field to a time in the past (we set it to the
epoch). Then it uses the Cookie Store API set function to set
this cookie so that the CFNetwork will update the existing cookies
with these options and delete them since they are expired.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_delete.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_multiple.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_delete_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::remove):
* Source/WebCore/Modules/cookie-store/CookieStore.h:

Canonical link: <a href="https://commits.webkit.org/266078@main">https://commits.webkit.org/266078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a29931f3b3bf5676116d32cf511449d121b8e2f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12781 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14520 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12212 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15609 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13127 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14912 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12946 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10818 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14969 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10970 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18641 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12045 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14935 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10102 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11450 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3138 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15764 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12031 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -192,14 +192,36 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+    CookieInit initOptions;
+    initOptions.name = WTFMove(options.name);
+    initOptions.value = ""_s;
+    initOptions.expires = 0;
```
Doesn't this end up creating a negative value in the dictionary we send to CFNetwork. I think we should use `WallTime::now() - 24h`

RupinMittal replied:
I think you're right. It may become negative when converted to a NSHTTPCookie. I can try `WallTime::now() - 24h`. The spec says that although the time doesn't matter as long as it's in the past, ideally we'd like to set it to the earliest representable date (https://wicg.github.io/cookie-store/#delete-a-cookie). Is there a good way to do this or do we prefer to just use `WallTime::now() - 24h`?

cdumez replied:
I stand by recommendation because we don't seem to have a good way to get the earliest representable date. Also, it doesn't matter as long as it is in the past.

RupinMittal replied:
Sounds good!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -192,14 +192,36 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        return;
+    }
+
+    CookieInit initOptions;
```
This should forward the CookieStoreDeleteOptions's domain/path to initOptions too.

RupinMittal replied:
That's what I had initially, but I excluded them because the spec says to use `null` and `'/'` for the domain and path when running delete (https://wicg.github.io/cookie-store/#CookieStore-delete). So I just let domain and path be default initialized to these values. Should I change this?

cdumez replied:
I see this in the spec:
```
Let r be the result of running [delete a cookie](https://wicg.github.io/cookie-store/#delete-a-cookie) with url, options["[name](https://wicg.github.io/cookie-store/#dom-cookiestoredeleteoptions-name)"], options["[domain](https://wicg.github.io/cookie-store/#dom-cookiestoredeleteoptions-domain)"], and options["[path](https://wicg.github.io/cookie-store/#dom-cookiestoredeleteoptions-path)"].
```

cdumez replied:
`null` and `/` are only for the delete() that takes a name, not for the delete that takes options as far as I can tell.

RupinMittal replied:
Ok that makes sense.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -192,14 +193,38 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+
+    CookieInit initOptions;
+    initOptions.name = WTFMove(options.name);
+    initOptions.value = ""_s;
```
`initOptions.value = emptyString();` is slightly more efficient.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -192,14 +193,38 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+    initOptions.value = ""_s;
+    initOptions.domain = WTFMove(options.domain);
+    initOptions.path = WTFMove(options.path);
+    initOptions.expires = WallTime::now().secondsSinceEpoch().milliseconds() - (24_h).milliseconds();
```
Couldn't you do this?
```
initOptions.expires = (WallTime::now() - 24_h).secondsSinceEpoch().milliseconds();
```

cdumez reviewed (approved):
LGTM with nits.

RupinMittal committed:
Implement the delete function in the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259220

Reviewed by Chris Dumez.

The delete function creates a cookie with the options given and
sets the expires field to a time in the past (we set it to the
epoch). Then it uses the Cookie Store API set function to set
this cookie so that the CFNetwork will update the existing cookies
with these options and delete them since they are expired.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_delete.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_multiple.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_delete_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_basic.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::remove):
* Source/WebCore/Modules/cookie-store/CookieStore.h:

Canonical link: https://commits.webkit.org/266078@main
